public/app/features/trails/utils.ts

```diff
@@ -175,12 +205,73 @@ export function limitAdhocProviders(
       // remove current selected filter if updating a chosen filter
       const filters = filtersValues.filter((f) => f.key !== filter.key);
       // call getTagValues and truncate the response
-      const values = (await datasourceHelper.getTagValues({ key: filter.key, filters })).slice(
-        0,
-        MAX_ADHOC_VARIABLE_OPTIONS
-      );
+      // we're passing the queries so we get the label values that adhere to the queries
+      // we're also passing the scopes so we get the label values that adhere to the scopes filters
+
+      const opts = {
+        key: filter.key,
+        filters,
+        scopes: getClosestScopesFacade(variable)?.value,
+        queries: dataTrail.getQueries(),
+      };
+
+      // if there are too many queries it takes to much time to process the requests.
+      // In this case we favour responsiveness over reducing the number of options.
+      if (opts.queries.length > 20) {
+        opts.queries = [];
+      }
+
+      const values = (await datasourceHelper.getTagValues(opts)).slice(0, MAX_ADHOC_VARIABLE_OPTIONS);
       // use replace: true to override the default lookup in adhoc filter variable
       return { replace: true, values };
     },
   });
 }
+
+export type SuggestionsResponse = {
+  data: string[];
+  status: 'success' | 'error';
+  error?: 'string';
+  warnings?: string[];
+};
+
+// Suggestions API is an API that receives adhoc filters, scopes and queries and returns the labels or label values that match the provided parameters
+// Under the hood it does exactly what the label and label values API where doing but the processing is done in the BE rather than in the FE
+export async function callSuggestionsApi(
+  dataSourceUid: string,
+  timeRange: RawTimeRange,
+  scopes: Scope[],
+  adHocVariableFilters: AdHocVariableFilter[],
+  labelName: string | undefined,
+  limit: number | undefined,
+  requestId: string
+): Promise<FetchResponse<SuggestionsResponse>> {
+  return await lastValueFrom(
+    getBackendSrv().fetch<SuggestionsResponse>({
+      url: `/api/datasources/uid/${dataSourceUid}/resources/suggestions`,
```

Comment:
Is there a case where we don't have suggestion API not available? If so, should we handle the errors?